Neos.Flow/Classes/Configuration/ConfigurationManager.php

```diff
@@ -188,7 +196,21 @@ public function __construct(ApplicationContext $context)
         $orderedListOfContextNames = [];
         $currentContext = $context;
         do {
-            $orderedListOfContextNames[] = (string)$currentContext;
+            if ($currentContext->getParent() !== null) {
+                $contextNameParts = explode('/', (string)$currentContext);
+                $n = count($contextNameParts) - 1;
+                // traverse all permutations of a context
+                for ($i = 0; $i < 2 ** $n; $i++) {
+                    $pattern = sprintf('%0' . $n . 's', decbin($i));
```

Comment:
This is really "clever" code, but it took me quite a while to understand what you are actually doing here and why that works. We should reconsider this and make it easier to grasp or, if not possible, clearly document, why this method to encode the permutation orders is used.

Basically, what you do is:
- Convert the permutation number (0 - 2^n) to binary with n digits
- replace every j-th context part with the wildcard, if the j-th bit is set

This resolves to following ordered list of contexts, given n = 3 (permutations = 2^3 = 8):
`000` => `Foo\Bar\Baz`
`001` => `Foo\Bar\Any`
`010` => `Foo\Any\Baz`
`011` => `Foo\Any\Any`
`100` => `Any\Bar\Baz`
`101` => `Any\Bar\Any`
`110` => `Any\Any\Baz`
`111` => `Any\Any\Any`

I'm unsure about the order in this case tbh., because that means that e.g. `Any\Bar\Any` is loaded after `Any\Bar\Baz`, no? Is that intended? Of course that is probably a rare edge case, since you most likely only have one or at most two sub-context levels, so n <= 2.

Also, if we stick with this, why not just bitmask instead of converting the bits to string and iterating the string?
```php
$wildcardedParts = array_map(function($index, $value) use ($i, $n) {
   // https://en.wikipedia.org/wiki/Mask_%28computing%29#Querying_the_status_of_a_bit
   $bitAtPositionIsSet = ($n & (0x1 << $index)) > 0;
   return $bitAtPositionIsSet ? self::CONFIGURATION_CONTEXT_WILDCARD : $value;
}, array_keys($contextNameParts), $contextNameParts);
```

Reply:
In the end of the initialization the completed array is - as it was before - [reversed](https://github.com/neos/flow-development-collection/blob/376177287cf0295cb79fb9689c3084f6481b1404/Neos.Flow/Classes/Configuration/ConfigurationManager.php#L215), so that should work as intended, but I probably should add a test for more complex subcontexts.
Using the binary counting for generating the permutations seemed to be more simple than a some (probably) recursive looping. Admittedly it's not really obvious.
I do however like the bitmasking idea, which I really should have thought of in the first place...